### PR TITLE
[core][Android] Fix slow boot time

### LINKED ISSUE #23189: [expo-modules][Android] - Single expo-modules library increases Android boot time by 2-2.5s
### Minimal reproducible example

https://github.com/intergalacticspacehighway/android-perf-test

### Summary

We have been debugging the slow startup performance of our Android app. Essentially, we maintained a single View and uninstalled numerous packages. Eventually, we discovered that the use of just a single library with expo-modules is enough to cause a boot delay on Android of at least _2-2.5 seconds._ **No impact on iOS**

How to reproduce:

Checkout the repo

```bash
yarn && yarn run android --variant Release
```

Measure the boot, then remove expo-blur and generate a new release. The startup performance should improve after removing expo-blur. The issue appears to be coming from the new expo modules API, as similar problems have been observed with other expo-* packages as well

### Environment

```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.14.2 - ~/.nvm/versions/node/v18.14.2/bin/node
      Yarn: 3.5.0 - ~/.nvm/versions/node/v18.14.2/bin/yarn
      npm: 9.5.0 - ~/.nvm/versions/node/v18.14.2/bin/npm
      Watchman: 2022.03.21.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.11.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    IDEs:
      Android Studio: 2022.1 AI-221.6008.13.2211.9514443
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmGlobalPackages:
      eas-cli: 3.8.1
      expo-cli: 6.3.2
    Expo Workflow: bare
```

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Closes https://github.com/expo/expo/issues/23189.

# How

Introducing a `LazyKType` which defers creation of the `KType` in time. 

# Test Plan

- bare-expo ✅
- [intergalacticspacehighway/android-perf-test](https://github.com/intergalacticspacehighway/android-perf-test) ✅

### CODE REVIEW

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt [hunk 1]
```diff
@@ -175,7 +184,7 @@ object TypeConverterProviderImpl : TypeConverterProvider {
     }
   }
 
-  private fun createCashedConverters(isOptional: Boolean): Map<KType, TypeConverter<*>> {
+  private fun createCashedConverters(isOptional: Boolean): Map<KClass<*>, TypeConverter<*>> {
```
Comment (alanjhughes) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt:187]:
```suggestion
  private fun createCachedConverters(isOptional: Boolean): Map<KClass<*>, TypeConverter<*>> {
```

**Review by tsapeta — APPROVED** [review]
🏎️

**Review by Kudo — APPROVED** [review]
🚀🚀🚀

**Comment by hirbod** [comment]
Thank you so much guys!

**Comment by RichardLindhout** [comment]
Is 48 also affected, if yes, will there be an update to 48 as well?

**Pull request merged**